lib/Dialect/Arc/ArcFolds.cpp

```diff
@@ -122,3 +122,137 @@ LogicalResult MemoryWriteOp::canonicalize(MemoryWriteOp op,
     return rewriter.eraseOp(op), success();
   return failure();
 }
+
+//===----------------------------------------------------------------------===//
+// ClockDomainOp
+//===----------------------------------------------------------------------===//
+
+static bool removeUnusedClockDomainInputs(ClockDomainOp op,
+                                          PatternRewriter &rewriter) {
+  bool changed = false;
+  for (auto arg :
+       llvm::make_early_inc_range(op.getBodyBlock().getArguments())) {
+    if (arg.use_empty()) {
+      auto i = arg.getArgNumber();
+      op.getBodyBlock().eraseArgument(i);
+      op.getInputsMutable().erase(i);
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+static bool removeUnusedClockDomainOutputs(ClockDomainOp op,
+                                           PatternRewriter &rewriter) {
+  SmallVector<Type> resultTypes;
+  for (auto res : llvm::reverse(op->getResults())) {
+    if (res.use_empty())
+      op.getBodyBlock().getTerminator()->eraseOperand(res.getResultNumber());
+    else
+      resultTypes.push_back(res.getType());
+  }
+
+  // Nothing is changed.
+  if (resultTypes.size() == op->getNumResults())
+    return false;
+
+  rewriter.setInsertionPoint(op);
+
+  auto newDomain = rewriter.create<ClockDomainOp>(
+      op.getLoc(), resultTypes, op.getInputs(), op.getClock());
+  rewriter.inlineRegionBefore(op.getBody(), newDomain.getBody(),
+                              newDomain->getRegion(0).begin());
+
+  unsigned currIdx = 0;
+  for (auto result : op.getOutputs()) {
+    if (!result.use_empty())
+      rewriter.replaceAllUsesWith(result, newDomain->getResult(currIdx++));
+  }
+
+  rewriter.eraseOp(op);
+  return true;
+}
+
+LogicalResult ClockDomainOp::canonicalize(ClockDomainOp op,
+                                          PatternRewriter &rewriter) {
+  rewriter.setInsertionPointToStart(&op.getBodyBlock());
+
+  // Canonicalize inputs
+  DenseMap<Value, unsigned> seenArgs;
+  for (auto arg :
+       llvm::make_early_inc_range(op.getBodyBlock().getArguments())) {
+    auto i = arg.getArgNumber();
+    if (arg.use_empty())
+      continue;
+
+    // Remove duplicate inputs
+    if (seenArgs.count(op.getInputs()[i])) {
```

Comment:
Nit:
```suggestion
    if (seenArgs.count(inputVal)) {
```
if you move the `inputVal` decl up to after `auto i = ...`